[[ Bug ]] Fix crash and behavior of popup widget

montegoulding opened the pull request:
This patch resolves a crash of popup widget when called in the absence of
a valid mouse stack. `MCWidgetPopupAtLocationWithProperties` no longer
translates the location parameter to a global location based on the mouse
stack. All translation is done prior to the call.

- When called from LCB the point is now translated relative to the host
widget stack which is what it is documented to do.
- When called from LCS the translation is relative to the target. This
allows for non-mouse events to call `popup widget`. For example, tabbing
a text field into focus. If no location was specified in LCS then the
mouse location is used and therefore translation is relative to the mouse
stack. Execution errors are now thrown if it is not possible to obtain a
valid global location.

montegoulding commented on engine/src/exec-interface.cpp:
```diff
@@ -2722,9 +2722,25 @@ void MCInterfaceExecPopupWidget(MCExecContext &ctxt, MCNameRef p_kind, MCPoint *
 	if (p_at != nil)
-		t_at = *p_at;
+    {
+        if (!MCtargetptr)
```
@runrevmark I guess there's a chance of someone doing this:
- not handling a message in the target but somewhere further down
- deleting the target
- popping up a widget

Should we permit that?

runrevmark replied:
I think the current behavior of throwing an error is fine if there is no target - after all the script specifies a relative location which is now relative to the target, so if there is no target object anymore, then that doesn['t make any sense.

montegoulding replied:
ok

montegoulding commented on engine/src/widget-popup.cpp:
```diff
@@ -426,9 +423,12 @@ extern "C" MC_DLLEXPORT_DEF MCValueRef MCWidgetExecPopupAtLocationWithProperties
+    t_host = MCWidgetGetHost(MCcurrentwidget);
+    
+    MCPoint t_at;
+	t_at = t_host->getstack()->globaltostackloc(MCGPointToMCPoint(MCWidgetMapPointToGlobal(MCcurrentwidget, t_point)));
```
@runrevmark I'm wondering if we might need to ensure `t_host->getstack()` is open and visible here or if there might be valid reasons for `popup widget` to be called in LCB when the host is not on screen.

runrevmark replied:
Hmm - another issue here is that we are assuming there is a 'current widget' - which precludes an LCB library from using popup widget :|

runrevmark replied:
Ah - sorry I just noticed that the whole call is guarded by EnsureCurrentWidget :)

Okay so I think we do need to ensure that the widget has an open and visible host - otherwise the 'at' relative position makes no sense.

montegoulding replied:
Hmm... the relative position still makes sense if the widget is not visible doesn't it?

runrevmark replied:
As mentioned offline, I meant to say 'open and visible host stack' :)

montegoulding committed:
[[ Bug ]] Fix crash and behavior of popup widget

This patch resolves a crash of popup widget when called in the absence of
a valid mouse stack. `MCWidgetPopupAtLocationWithProperties` no longer
translates the location parameter to a global location based on the mouse
stack. All translation is done prior to the call.

- When called from LCB the point is now translated relative to the host
widget stack which is what it is documented to do.
- When called from LCS the translation is relative to the target. This
allows for non-mouse events to call `popup widget`. For example, tabbing
a text field into focus. If no location was specified in LCS then the
mouse location is used and therefore translation is relative to the mouse
stack. Execution errors are now thrown if it is not possible to obtain a
valid global location.

runrevmark commented:
@livecode-vulcan review ok 1993933

livecode-vulcan commented:
💙 review by @runrevmark ok 1993933db7706dd0828b55a367361345d818638a

livecode-vulcan commented:
😎 test success 1993933db7706dd0828b55a367361345d818638a

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/489)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/481)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/481)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/481)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/481)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2003)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/231)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/230)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2035)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2025)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2023)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2058)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1995)

runrevmark commented:
@montegoulding This patch has caused a regression - https://quality.livecode.com/show_bug.cgi?id=22603 - "When called from LCB the point is now translated relative to the host
widget stack which is what it is documented to do." isn't actually how it is documented in LCB - its relative to the top-left of the widget. The reason here is that all co-ord related things fetched widget methods tend to relative to the widget - i.e. the click position is relative to the top-left of the widget.
